### PR TITLE
Add prometheus cluster monitoring addon.

### PR DESCRIPTION
**Pull request by serathius**
This PR adds new cluster monitoring addon based on prometheus.
It adds prometheus deployment with e2e tests.
Additional components will be added iterativly in future.
Manifests based on current Helm chart.
At current state it's not intended for production use.

cc @piosz @kawych @miekg
```release-note
Add prometheus cluster monitoring addon to kube-up
```
/sig instrumentation
/kind feature
/priority important-soon

### CODE REVIEW

File: cluster/gce/gci/configure-helper.sh [hunk 1]
```diff
@@ -2104,14 +2104,17 @@ EOF
     prepare-kube-proxy-manifest-variables "$src_dir/kube-proxy/kube-proxy-ds.yaml"
     setup-addon-manifests "addons" "kube-proxy"
   fi
+  if [[ "${ENABLE_CLUSTER_MONITORING:-}" != "none" ]]; then
```
Comment (kawych) [inline cluster/gce/gci/configure-helper.sh]: I prefer `if [[ "${ENABLE_CLUSTER_MONITORING:-}" == "prometheus" ]]`. It makes sense to have fully separated conditions for prometheus and other monitoring systems, because this is the only one that doesn't use heapster. Please add some comments to make this separation clear, e.g. "set up cluster monitoring using prometheus" and "set up cluster monitoring using heapster"

File: cluster/addons/cluster-monitoring/prometheus/prometheus-pvc.yaml [hunk 2]
```diff
@@ -0,0 +1,16 @@
+---
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  name: prometheus
+  namespace: kube-system
+  labels:
+    kubernetes.io/cluster-service: "true"
+    addonmanager.kubernetes.io/mode: EnsureExists
```
Comment (kawych) [inline cluster/addons/cluster-monitoring/prometheus/prometheus-pvc.yaml]: Do we actually intend users to modify this, i.e. other parts than storage request?

File: cluster/addons/cluster-monitoring/prometheus/prometheus-configmap.yaml [hunk 3]
```diff
@@ -0,0 +1,190 @@
+---
+apiVersion: v1
+kind: ConfigMap
```
Comment (kawych) [inline cluster/addons/cluster-monitoring/prometheus/prometheus-configmap.yaml]: Can you include some reference for the format of this config map?

File: cluster/addons/cluster-monitoring/prometheus/prometheus-configmap.yaml [hunk 4]
```diff
@@ -0,0 +1,190 @@
+---
```
Comment (kawych) [inline cluster/addons/cluster-monitoring/prometheus/prometheus-configmap.yaml]: nit: please skip unnecessary separator lines like this (all first lines)

**Comment by kawych** [comment]
The deployments look fine, I'll take a look at the e2e tests on Monday. Can you split this PR to two commits: deplyments and tests?

File: cluster/addons/cluster-monitoring/prometheus/prometheus-deployment.yaml [hunk 5]
```diff
@@ -0,0 +1,86 @@
+---
+apiVersion: extensions/v1beta1
+kind: Deployment
```
Comment (kawych) [inline cluster/addons/cluster-monitoring/prometheus/prometheus-deployment.yaml]: Do you know what is prometheus cpu/memory usage and whether we can rely on defaults?

File: cluster/addons/cluster-monitoring/prometheus/prometheus-configmap.yaml [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+    - job_name: kubernetes-nodes-kubelet
+      kubernetes_sd_configs:
+      - role: node
+      relabel_configs:
+      - action: labelmap
+        regex: __meta_kubernetes_node_label_(.+)
+      - replacement: kubernetes.default.svc:443
+        target_label: __address__
+      - regex: (.+)
+        replacement: /api/v1/nodes/${1}/proxy/metrics
```
Comment (brancz) [inline cluster/addons/cluster-monitoring/prometheus/prometheus-configmap.yaml]: I'm not sure this is a good idea to advocate for to users. People will look at this and think this is the recommended way to run this, but in reality it's giving close to root access to the Prometheus pod to all kubelets, that doesn't seem like a good idea. I would prefer cert or token based authN + authZ from the kubelet.

Comment (serathius) [reply cluster/addons/cluster-monitoring/prometheus/prometheus-configmap.yaml]: Looks like kube-up has http endpoints enabled for kubelet. I will use it as temporary solution and work on solution in mean time.

File: test/utils/runners.go [hunk 7]
```diff
@@ -127,6 +127,7 @@ type RCConfig struct {
 	ReadinessProbe    *v1.Probe
 	DNSPolicy         *v1.DNSPolicy
 	PriorityClassName string
+	PodAnnotations    map[string]string
```
Comment (kawych) [inline test/utils/runners.go]: nit: this is similar to labels, so probably it would fit better after Labels field

File: test/e2e/instrumentation/monitoring/prometheus.go [hunk 8]
```diff
@@ -0,0 +1,385 @@
...
+	for {
+		if predicate() {
+			return
+		}
+		if time.Since(startTime) >= testTimeout {
+			break
+		}
+		time.Sleep(prometheusSleepBetweenAttempts)
+	}
+	framework.Failf("monitoring using prometheus test failed")
```
Comment (kawych) [inline test/e2e/instrumentation/monitoring/prometheus.go]: Can you make the error message more useful, e.g. number of nodes with metrics vs all nodes, actual metric value vs expected, etc?

File: test/e2e/instrumentation/monitoring/prometheus.go [hunk 9]
```diff
@@ -0,0 +1,385 @@
...
+	}
+
+	if !expectedItemsExist(expectedNodesNames, nodesUp) {
+		framework.Logf("Failed to find all expected Nodes.\n Expected: %v\nActual: %v", expectedNodesNames, nodesUp)
+		return false
+	}
+	return true
+}
+
+func validateQueryValues(c clientset.Interface, query string, expectedValue float64, samplesCount int, errorTolerance float64) bool {
```
Comment (kawych) [inline test/e2e/instrumentation/monitoring/prometheus.go]: nit: please stick to some naming convention, e.g. use "validate" prefix in all functions that check if the result is expected or use some other name here ("metricHasCorrectValue"). Personally I'd prefer the latter and use validate.* for functions that do some validation and fail the test/return error if it doesn't pass.

File: test/e2e/instrumentation/monitoring/prometheus.go [hunk 10]
```diff
@@ -0,0 +1,385 @@
...
+
+func allScrapeTargetsAreHealthy(c clientset.Interface) bool {
+	discovery, err := fetchPrometheusTargetDiscovery(c)
+	if err != nil {
+		framework.Logf("failed to query prometheus. Err: %v", err)
+		return false
+	}
+	if len(discovery.ActiveTargets) == 0 {
+		framework.Logf("prometheus is not scraping any targets")
+		return false
```
Comment (kawych) [inline test/e2e/instrumentation/monitoring/prometheus.go]: What targets are required?

File: test/e2e/instrumentation/monitoring/prometheus.go [hunk 11]
```diff
@@ -0,0 +1,385 @@
...
+	for _, target := range discovery.ActiveTargets {
+		if target.Health != HealthGood {
+			framework.Logf("target health not good. Target: %v", target)
+			return false
+		}
+	}
+	return true
+}
+
+func metricAvailableOnNodes(c clientset.Interface, expectedNodesNames []string, metric string, samplesCount int) bool {
```
Comment (kawych) [inline test/e2e/instrumentation/monitoring/prometheus.go]: s/metricAvailableOnNodes/metricAvailableOnAllNodes

File: test/e2e/instrumentation/monitoring/prometheus.go [hunk 12]
```diff
@@ -0,0 +1,385 @@
...
+	if err != nil {
+		framework.Logf("failed to query prometheus. Err: %v", err)
+		return false
+	}
+	if len(samples) < samplesCount {
+		framework.Logf("not enough samples, got %v", samples)
+		return false
+	}
+	framework.Logf("Received metric: %v", samples)
+	for _, value := range samples {
```
Comment (kawych) [inline test/e2e/instrumentation/monitoring/prometheus.go]: Do we expect more than one sample? If not, can you make it explicit?

File: test/e2e/instrumentation/monitoring/prometheus.go [hunk 13]
```diff
@@ -0,0 +1,389 @@
...
+		err = json.Unmarshal(v.Result, &mv)
+		qr.v = mv
+
+	default:
+		err = fmt.Errorf("unexpected value type %q", v.Type)
+	}
+	return err
+}
+
+func retryUntil(validator func() error, timeout time.Duration) {
```
Comment (kawych) [inline test/e2e/instrumentation/monitoring/prometheus.go]: nit: maybe retryUntilSucceeds, since you don't use a predicate anymore?

File: test/e2e/instrumentation/monitoring/prometheus.go [hunk 14]
```diff
@@ -0,0 +1,389 @@
...
+	return nil
+}
+
+func validateAllActiveTargetsAreHealthy(c clientset.Interface) error {
+	discovery, err := fetchPrometheusTargetDiscovery(c)
+	if err != nil {
+		return err
+	}
+	if len(discovery.ActiveTargets) == 0 {
+		return fmt.Errorf("Prometheus is not scraping any targets, at least one target is required")
```
Comment (kawych) [inline test/e2e/instrumentation/monitoring/prometheus.go]: What targets are required? I don't ask you do change code here, just want to understand it better. E.g. is a system pod an active target? Why do we expect all targets to be healthy, will it be reported unhealthy if some system component is failing? Should it cause Prometheus tests to fail?

File: test/e2e/instrumentation/monitoring/prometheus.go [hunk 15]
```diff
@@ -0,0 +1,389 @@
...
+		return err
+	}
+	nodesWithMetric := make(map[string]bool)
+	for _, instance := range instanceLabels {
+		nodesWithMetric[instance] = true
+	}
+
+	for _, nodeName := range expectedNodesNames {
+		if _, found := nodesWithMetric[nodeName]; !found {
+			return fmt.Errorf("Metric '%v' is not available for node '%v'", metric, nodeName)
```
Comment (kawych) [inline test/e2e/instrumentation/monitoring/prometheus.go]: I think the number of nodes would be more useful, for automatically running tests the cluster is deleted before anyone can ssh to a node and debug it. And information if the metrics are available for all nodes vs most nodes vs no nodes already gives some idea.

File: test/e2e/instrumentation/monitoring/prometheus.go [hunk 16]
```diff
@@ -0,0 +1,389 @@
...
+		serviceAnnotations = map[string]string{
+			"prometheus.io/scrape": "true",
+			"prometheus.io/path":   "/Metrics",
+			"prometheus.io/port":   "8080",
+		}
+	}
+	return common.NewMetricExporter(consumerName, f.Namespace.Name, podAnnotations, serviceAnnotations, metricValue, f.ClientSet, f.InternalClientset, f.ScalesGetter)
+}
+
+func validateMetricAvailableForNodes(c clientset.Interface, metric string, expectedNodesNames []string) error {
```
Comment (kawych) [inline test/e2e/instrumentation/monitoring/prometheus.go]: s/validateMetricAvailableForNodes/validateMetricAvailableForAllNodes

File: test/e2e/instrumentation/monitoring/prometheus.go [hunk 17]
```diff
@@ -0,0 +1,389 @@
...
+		}, prometheusTestTimeout)
+	})
+	It("should successfully scrape all targets", func() {
+		retryUntil(func() error {
+			return validateAllActiveTargetsAreHealthy(f.ClientSet)
+		}, prometheusTestTimeout)
+	})
+	It("should contain correct container CPU metric.", func() {
+		var targetCPUUsage float64 = 0.2
+		var requiredSampleCount int = 3
```
Comment (kawych) [inline test/e2e/instrumentation/monitoring/prometheus.go]: Why is it 3? Looking at this code, I would expect one resource consumer pod that consumes 200m CPU, where do the other samples come from?

File: test/e2e/instrumentation/monitoring/prometheus.go (hunk 14 above)
Comment (serathius) [reply test/e2e/instrumentation/monitoring/prometheus.go]: Target is url endpoint that is scraped by prometheus. It can be anything that is found by prometheus service discovery or hardcoded (nodes, pods, services, ingress, apiserver). Prometheus applies rules on those targets that can modify them, add labels or discard them. Active targets are those that were not discarded. For kubernetes we will drop all targets that do not have proper annotations. 

Not healthy targets can mean that prometheus didn't manage to run request (there is delay until requests are made or prometheus wasn't fast enough to make request) or request failed (misconfiguration, timeout, unauthorized).

In that test I want mainly verify configuration, that all targets are reachable.

File: test/e2e/instrumentation/monitoring/prometheus.go (hunk 13 above)
Comment (serathius) [reply test/e2e/instrumentation/monitoring/prometheus.go]: ok, makes more sense.

File: test/e2e/instrumentation/monitoring/prometheus.go (hunk 15 above)
Comment (serathius) [reply test/e2e/instrumentation/monitoring/prometheus.go]: ok, "3 out of 5 nodes"

File: test/e2e/instrumentation/monitoring/prometheus.go (hunk 17 above)
Comment (serathius) [reply test/e2e/instrumentation/monitoring/prometheus.go]: Sample is generated every 2 minutes (based on rate passed to query) for same pod.
Based on how cpu metric is calculated, value will linearly grow until reaches expected value. To ensure that value has stabilized I use 3 samples.

File: test/e2e/instrumentation/monitoring/prometheus.go (hunk 16 above)
Comment (serathius) [reply test/e2e/instrumentation/monitoring/prometheus.go]: ok

File: test/e2e/instrumentation/monitoring/prometheus.go (hunk 14 above)
Comment (kawych) [reply test/e2e/instrumentation/monitoring/prometheus.go]: Got it, this is very informative, consider adding this explanation (or part of it) in the code comments as well.

I'm also thinking if it's possible to hit a window when only one or few targets are discovered and are succeeding, and we didn't get to test some failing targets. Do you think it makes sense to wait for some more representative set of active targets to test? (Not necessarily in this PR)

File: test/e2e/instrumentation/monitoring/prometheus.go [hunk 18]
```diff
@@ -0,0 +1,393 @@
...
+		var annotatePods, annotateService bool = true, false
+		query := prometheusPodCustomMetricQuery(f.Namespace.Name, consumerName)
+		consumer := exportCustomMetric(f, consumerName, int(customMetricTargetValue), annotatePods, annotateService)
+		defer consumer.CleanUp()
+		retryUntilSucceeds(func() error {
+			return validateQueryReturnsCorrectValues(f.ClientSet, query, customMetricTargetValue, requiredSampleCount, prometheusMetricErrorTolerance)
+		}, prometheusTestTimeout)
+	})
+	It("should scrape metrics from annotated services.", func() {
+		var customMetricTargetValue float64 = 1000
```
Comment (kawych) [inline test/e2e/instrumentation/monitoring/prometheus.go]: I think you missed my previous comment - can you use lliterals instead? (you can use bool constants for readability: doAnnotatePods = true, doAnnotateServices = true, dontAnnotatePods = false, dontAnnotateServices = false)

File: cluster/gce/gci/configure-helper.sh [hunk 19]
```diff
@@ -2104,6 +2104,12 @@ EOF
     prepare-kube-proxy-manifest-variables "$src_dir/kube-proxy/kube-proxy-ds.yaml"
     setup-addon-manifests "addons" "kube-proxy"
   fi
+  # Setup cluster monitoring using prometheus
```
Comment (kawych) [inline cluster/gce/gci/configure-helper.sh]: AFAIU we may still need to run standalone Heapster in addition to this - for users that use older versions of kubectl

File: test/e2e/instrumentation/monitoring/prometheus.go (hunk 18 above)
Comment (serathius) [reply test/e2e/instrumentation/monitoring/prometheus.go]: I moved test parameterization to consts and split `exposeCustomMetrics` function based on annotation type.

File: cluster/gce/gci/configure-helper.sh (hunk 19 above)
Comment (serathius) [reply cluster/gce/gci/configure-helper.sh]: I will remove prometheus from cluster monitoring addons to allow deploy heapster separately.

File: test/e2e/instrumentation/monitoring/prometheus.go (hunk 14 above)
Comment (serathius) [reply test/e2e/instrumentation/monitoring/prometheus.go]: Some targets are available from configuration (static prometheus self monitoring) and other are depended on service discovery that requires call to kubernetes apiserver. I would guess that there is window between reloading configuration and apiserver response (especially for large clusters).

We don't want to pin to specific job, but we may want to ensure that targets were discovered and will not change.

File: cluster/gce/gci/configure-helper.sh [hunk 20]
```diff
@@ -2104,6 +2104,11 @@ EOF
     prepare-kube-proxy-manifest-variables "$src_dir/kube-proxy/kube-proxy-ds.yaml"
     setup-addon-manifests "addons" "kube-proxy"
   fi
+  # Setup prometheus stack for monitoring kubernetes cluster
+  if [[ "${ENABLE_PROMETHEUS_MONITORING:-}" == "true" ]]; then
```
Comment (kawych) [inline cluster/gce/gci/configure-helper.sh]: Have you tested it? If you add new env var, I think you need to add it here as well: https://github.com/kubernetes/kubernetes/blob/master/cluster/gce/util.sh#L678

File: test/e2e/instrumentation/monitoring/prometheus.go (hunk 14 above)
Comment (kawych) [reply test/e2e/instrumentation/monitoring/prometheus.go]: Sounds good, something like checking that there is the same number of targets between two iterations before proceeding?

**Comment by kawych** [comment]
LGTM. Please fix presubmits and make sure that other reviewers approve (or uncc them).

File: cluster/gce/gci/configure-helper.sh (hunk 20 above)
Comment (serathius) [reply cluster/gce/gci/configure-helper.sh]: Just finished testing it from fresh cluster. Moving prometheus out of cluster-monitoring took more changes.

File: test/e2e/instrumentation/monitoring/prometheus.go (hunk 14 above)
Comment (serathius) [reply test/e2e/instrumentation/monitoring/prometheus.go]: I was thinking about something right that with small percent tolerations to handle large clusters.

File: cluster/addons/cluster-monitoring/prometheus/prometheus-configmap.yaml (hunk 6 above)
Comment (serathius) [reply cluster/addons/cluster-monitoring/prometheus/prometheus-configmap.yaml]: @brancz Is using unencrypted metric endpoints acceptable for first version? I plan to support this addon through changes into kubelet metrics. For this PR I wanted to move current community solution for prometheus into addon, but enhance it with e2e tests.

File: cluster/addons/prometheus/prometheus-configmap.yaml [hunk 21]
```diff
@@ -0,0 +1,176 @@
...
+  namespace: kube-system 
+  labels:
+    kubernetes.io/cluster-service: "true"
+    addonmanager.kubernetes.io/mode: EnsureExists
+data:
+  alerts: |
+    {}
+  prometheus.yml: |
+    rule_files:
+    - /etc/config/rules
```
Comment (brancz) [inline cluster/addons/prometheus/prometheus-configmap.yaml]: why specify rules and alert files if they are then left empty?

File: cluster/addons/prometheus/prometheus-deployment.yaml [hunk 22]
```diff
@@ -0,0 +1,100 @@
...
+          resources:
+            limits:
+              cpu: 10m
+              memory: 10Mi
+            requests:
+              cpu: 10m
+              memory: 10Mi
+
+        - name: prometheus-server
+          image: "prom/prometheus:v2.1.0"
```
Comment (brancz) [inline cluster/addons/prometheus/prometheus-deployment.yaml]: v2.1.0 had a variety of problems, I'd recommend v2.2.1

**Review by brancz — COMMENTED** [review]
generally looks good, just two suggestions

**Comment by brancz** [comment]
What I'm trying to understand is do we exclusively want to use this for the e2e tests of custom metrics or promote this as an official addon? If the latter then I'm not sure I'm comfortable with this. For testing purposes the insecure port is totally fine, but in production environments it's not what I would recommend.

Personally I'd of course like to see this done with the Prometheus Operator :wink: .

File: cluster/addons/prometheus/prometheus-configmap.yaml (hunk 21 above)
Comment (serathius) [reply cluster/addons/prometheus/prometheus-configmap.yaml]: Removed

File: cluster/addons/prometheus/prometheus-deployment.yaml (hunk 22 above)
Comment (serathius) [reply cluster/addons/prometheus/prometheus-deployment.yaml]: Updated

**Comment by brancz** [comment]
The Prometheus Operator is soon going to implement the TokenRequest API, in order to use tokens for specific audiences. As far as I understand token impersonation is why token auth on kubelets is not enabled by GKE today (RE: https://github.com/kubernetes/kubernetes/pull/57997). So until TokenRequests are available it won't be possible on GKE. This is somewhat reasonable I guess (although personally I feel the kubelet is higher privileged than Prometheus, so impersonation would not be a security concern, but I don't want to start that discussion here, also I'm happy to be proven wrong, I admit I haven't analyzed the security situation to its fullest).

Eventually I'd prefer to see this Prometheus Operator based as it solves a lot of operational needs of Prometheus (the TokenRequest being only one example, which is [unlikely to land in Prometheus itself](https://github.com/prometheus/prometheus/issues/4088)). Also we maintain a rather exhaustive setup already to perform [cluster monitoring](https://github.com/coreos/prometheus-operator/tree/master/contrib/kube-prometheus), which we have [productionized on top of OpenShift](https://github.com/openshift/cluster-monitoring-operator) and are planning to add support for vanilla Kubernetes as well.

tl;dr I'm ok with this state for now, but I'd prefer if we don't commit to this in the long term as we already know of shortcomings of this and diverge to a Prometheus Operator based setup.

(disclaimer I'm one of the maintainers of the Prometheus Operator)

**Comment by kawych** [comment]
/lgtm
@brancz thank you for explanations. From my knowledge, token auth is going to be enabled (see #58178), as discussed with @serathius we can move away from insecure port in a follow-up PR.

@piosz and @serathius should be able to contribute more to discussion about using Prometheus Operator. I'm not fully aware of the benefits of Prometheus Operator that are already available, @serathius has been investigating this more, i.e. he raised a concern that we may need some wider review of Prometheus Operator CRDs.

**Comment by brancz** [comment]
I would like us to discuss in more detail what we want to achieve here. From the sig-instrumentation meeting two weeks ago it I was under the impression that all we wanted to do is a very simple setup purely to validate in e2e tests that Kubernetes SD and other integrations are not totally broken. A fully fledged cluster monitoring addon is another story. I would like us to reconsider this.
